main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,263 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+        (Expression.LoadBool(e, offset), b)
+      case Expression.LoadInt8(exp, offset) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.LoadInt8(e, offset), b)
+      case Expression.LoadInt16(exp, offset) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.LoadInt16(e, offset), b)
+      case Expression.LoadInt32(exp, offset) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.LoadInt32(e, offset), b)
+      case Expression.StoreBool(exp, offset, v) =>
+        val (e1, b1) = optimizeExp(exp)
+        val (e2, b2) = optimizeExp(v)
+        (Expression.StoreBool(e1, offset, e2), b1 || b2)
+      case Expression.StoreInt8(exp, offset, v) =>
+        val (e1, b1) = optimizeExp(exp)
+        val (e2, b2) = optimizeExp(v)
+        (Expression.StoreInt8(e1, offset, e2), b1 || b2)
+      case Expression.StoreInt16(exp, offset, v) =>
+        val (e1, b1) = optimizeExp(exp)
+        val (e2, b2) = optimizeExp(v)
+        (Expression.StoreInt16(e1, offset, e2), b1 || b2)
+      case Expression.StoreInt32(exp, offset, v) =>
+        val (e1, b1) = optimizeExp(exp)
+        val (e2, b2) = optimizeExp(v)
+        (Expression.StoreInt32(e1, offset, e2), b1 || b2)
+      case Expression.Var(sym, tpe, loc) => (Expression.Var(sym, tpe, loc), false)
+      case Expression.Ref(sym, tpe, loc) => (Expression.Ref(sym, tpe, loc), false)
+      case Expression.Lambda(args, body, tpe, loc) =>
+        val (e, b) = optimizeExp(body)
+        (Expression.Lambda(args, e, tpe, loc), b)
+      case Expression.Hook(hook, tpe, loc) => (Expression.Hook(hook, tpe, loc), false)
+      case Expression.MkClosure(lambda, freeVars, tpe, loc) =>
+        val (e, b) = optimizeExp(lambda)
+        (Expression.MkClosure(e.asInstanceOf[Expression.Lambda], freeVars, tpe, loc), b)
+      case Expression.MkClosureRef(ref, freeVars, tpe, loc) =>
+        val (e, b) = optimizeExp(ref)
+        (Expression.MkClosureRef(e.asInstanceOf[Expression.Ref], freeVars, tpe, loc), b)
+      case Expression.ApplyRef(sym, args, tpe, loc) =>
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.ApplyRef(sym, es, tpe, loc), bs.exists(identity))
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) =>
+        val (es, bs) = optimizeExps(actuals).unzip
+        (Expression.ApplyTail(sym, formals, es, tpe, loc), bs.exists(identity))
+      case Expression.ApplyHook(hook, args, tpe, loc) =>
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.ApplyHook(hook, es, tpe, loc), bs.exists(identity))
+      case Expression.Apply(exp, args, tpe, loc) =>
+        val (e, b) = optimizeExp(exp)
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.Apply(e, es, tpe, loc), bs.exists(identity) || b)
+      case Expression.Unary(op, exp, tpe, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Unary(op, e, tpe, loc), b)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) =>
```

Comment:
If the arguments are literals you can do constant folding, e.g. `4 + 4 = 8`. Same for unary expressions.